bookkeeper-server/src/main/java/org/apache/bookkeeper/client/MetadataUpdateLoop.java

```diff
@@ -0,0 +1,186 @@
+ * 4. Read the metadata back from the store
+ * 5. Update the local copy of the metadata with the metadata read in 4, go to 1.
+ * 6. Update the local copy of the metadata with the metadata which has just been written.
+ *
+ * <p>All mutating operations are compare and swap operation. If the compare fails, another
+ * iteration of the loop begins.
+ */
+class MetadataUpdateLoop {
+    static final Logger LOG = LoggerFactory.getLogger(MetadataUpdateLoop.class);
+
+    private final LedgerManager lm;
+    private final long ledgerId;
+    private final Supplier<LedgerMetadata> currentLocalValue;
+    private final NeedsUpdatePredicate needsTransformation;
+    private final Function<LedgerMetadata, LedgerMetadata> transform;
+    private final BiPredicate<LedgerMetadata, LedgerMetadata> updateLocalValue;
+
+    private final String logContext;
+    private final AtomicInteger writeLoopCounter = new AtomicInteger(0);
+
+    interface NeedsUpdatePredicate {
+        boolean needsUpdate(LedgerMetadata metadata) throws Exception;
+    }
+
+    /**
+     * Construct the loop. This takes a set of functions which may be called multiple times
+     * during the loop.
+     *
+     * @param currentLocalValue should return the current local value of the metadata.
+     * @param needsTransformation should return true, if the metadata needs to be modified.
+     *                            should throw an exception, if this update doesn't make sense.
+     * @param transform takes a metadata objects, transforms, and returns it, without modifying
+     *                  the original.
+     * @param updateLocalValue if the local value matches the first parameter, update it to the
+     *                         second parameter and return true, return false otherwise.
+     */
+    MetadataUpdateLoop(LedgerManager lm,
+                       long ledgerId,
+                       Supplier<LedgerMetadata> currentLocalValue,
+                       NeedsUpdatePredicate needsTransformation,
+                       Function<LedgerMetadata, LedgerMetadata> transform,
+                       BiPredicate<LedgerMetadata, LedgerMetadata> updateLocalValue) {
+        this.lm = lm;
+        this.ledgerId = ledgerId;
+        this.currentLocalValue = currentLocalValue;
+        this.needsTransformation = needsTransformation;
+        this.transform = transform;
+        this.updateLocalValue = updateLocalValue;
+
+        this.logContext = String.format("UpdateLoop(ledgerId=%d,loopId=%s)",
+                                        ledgerId, UUID.randomUUID().toString().substring(0, 8));
+    }
+
+    CompletableFuture<LedgerMetadata> run() {
+        CompletableFuture<LedgerMetadata> promise = new CompletableFuture<>();
+
+        writeLoop(currentLocalValue.get(), promise);
+
+        return promise;
+    }
+
+    private void writeLoop(LedgerMetadata currentLocal, CompletableFuture<LedgerMetadata> promise) {
```

Comment:
@sijie @eolivelli The patch appears bigger than it actually is. The meat of it is in this file, from here down.

Reply:
will take a look at it today